### PR TITLE
[RFC/WIP] improve zfs locking

### PR DESCRIPTION
**Pull request by brauner**

### CODE REVIEW

File: lxd/storage_zfs.go [hunk 1]
```diff
@@ -1249,18 +1251,13 @@ func (s *storageZfs) ContainerSnapshotCreateEmpty(snapshotContainer container) e
...
-// - mark new zfs volume images/<fingerprint> readonly
-// - remove mountpoint property from zfs volume images/<fingerprint>
-// - create read-write snapshot from zfs volume images/<fingerprint>
 func (s *storageZfs) ImageCreate(fingerprint string) error {
 	shared.LogDebugf("Creating ZFS storage volume for image \"%s\" on storage pool \"%s\".", fingerprint, s.pool.Name)
 
-	imageMntPoint := getImageMountPoint(s.pool.Name, fingerprint)
-	fs := fmt.Sprintf("images/%s", fingerprint)
+	poolName := s.getOnDiskPoolName()
+	imageBaseDataset := fmt.Sprintf("images/%s", fingerprint)
```
Comment (stgraber) [inline lxd/storage_zfs.go]: Inconsistent naming with other function above.

File: lxd/storage.go [hunk 2]
```diff
@@ -27,6 +27,16 @@ import (
...
 
+type containerCreationMap struct {
+	refcount map[string]int
+	wait     map[string]chan bool
+}
+
+var lxdContainerCreationMap = containerCreationMap{
+	refcount: make(map[string]int),
+	wait:     make(map[string]chan bool),
+}
```
Comment (stgraber) [inline lxd/storage.go]: Shouldn't that be under lock to prevent concurrent read/write to a shared map?

Comment (stgraber) [inline lxd/storage.go]: Oh, I see, you're using lxdStorageMapLock

File: lxd/storage.go [hunk 3]
```diff
@@ -27,6 +27,16 @@ import (
 // Note that any access to this map must be done while holding a lock.
 var lxdStorageOngoingOperationMap = map[string]chan bool{}
 
+type containerCreationMap struct {
+	refcount map[string]int
```
Comment (stgraber) [inline lxd/storage.go]: That kind of handmade refcouting is a bit weird in Go, why not using sync.WaitGroup as we are doing in a bunch of other places?

Comment (brauner) [inline lxd/storage.go]: Yeah, I might just switch to that.

**Comment by brauner** [comment]
There's a bunch of other intricacies I need to think about before proceeding with this. But zfs will be the first target.

File: lxd/storage_zfs.go [hunk 4]
```diff
@@ -542,42 +543,45 @@ func (s *storageZfs) ContainerCreate(container container) error {
...
+	lxdContainerCreationLock.Unlock()
 
-		var imgerr error
-		if !s.zfsFilesystemEntityExists(fsImage, true) {
-			imgerr = s.ImageCreate(fingerprint)
-		}
+	// ImageCreate() does not need to and cannot be placed under the
+	// container lock. Otherwise we'll deadlock.
+	fsImage := fmt.Sprintf("images/%s", fingerprint)
+	err := s.ImageCreate(fingerprint)
```
Comment (brauner) [inline lxd/storage_zfs.go]: Note: Open to race, `ImageDelete()` might grab a lock right after and delete the image.

**Comment by stgraber** [comment]
Closing this one for now.

**Pull request closed without merging**